File: app/src/main/java/com/mehmetakiftutuncu/muezzin/activities/PrayerTimesActivity.java
package com.mehmetakiftutuncu.muezzin.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.github.mehmetakiftutuncu.toolbelt.Optional;
import com.mehmetakiftutuncu.muezzin.R;
import com.mehmetakiftutuncu.muezzin.activities.preferences.PreferencesActivity;
import com.mehmetakiftutuncu.muezzin.fragments.NoPlacesFoundFragment;
import com.mehmetakiftutuncu.muezzin.fragments.PrayerTimesFragment;
import com.mehmetakiftutuncu.muezzin.models.Place;
import com.mehmetakiftutuncu.muezzin.utilities.Pref;
import com.stephentuso.welcome.WelcomeScreenHelper;

public class PrayerTimesActivity extends MuezzinActivity {
    private WelcomeScreenHelper welcomeScreenHelper;
    private boolean shownWelcomeScreen;

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_prayertimes);

        welcomeScreenHelper = new WelcomeScreenHelper(this, WelcomeActivity.class);

        shownWelcomeScreen = savedInstanceState != null && savedInstanceState.getBoolean("shownWelcomeScreen", false);

        if (!shownWelcomeScreen) {
            shownWelcomeScreen = true;

            if (Pref.Application.getVersion(this) < 4) {
                // Installed version 2.0 for the first time
                Toast.makeText(this, R.string.welcome_updateNotice, Toast.LENGTH_LONG).show();
                Pref.getSharedPreferences(this).edit().clear().apply();
                welcomeScreenHelper.forceShow();
                Pref.Application.setVersion(this);
            } else {
                welcomeScreenHelper.show(savedInstanceState);
            }
        }
    }

    @Override protected void onResume() {
        super.onResume();

        Optional<Place> maybeCurrentPlace = Pref.Places.getCurrentPlace(this);

        if (maybeCurrentPlace.isEmpty()) {
            showNoPlacesFound();
        } else {
            Place currentPlace = maybeCurrentPlace.get();

            showPrayerTimes(currentPlace.toBundle());
        }
    }

    @Override public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.settings, menu);

        return super.onCreateOptionsMenu(menu);
    }

    @Override public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_preferences:
                startActivity(new Intent(this, PreferencesActivity.class));

                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override protected void onSaveInstanceState(Bundle outState) {
        welcomeScreenHelper.onSaveInstanceState(outState);
        outState.putBoolean("shownWelcomeScreen", shownWelcomeScreen);

        super.onSaveInstanceState(outState);
    }

    private void showNoPlacesFound() {
        NoPlacesFoundFragment noPlacesFoundFragment = new NoPlacesFoundFragment();

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.frameLayout_prayerTimesContainer, noPlacesFoundFragment, "NoPlacesFoundFragment")
                .commit();
    }

    private void showPrayerTimes(Bundle bundle) {
        PrayerTimesFragment prayerTimesFragment = PrayerTimesFragment.with(bundle);

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.frameLayout_prayerTimesContainer, prayerTimesFragment, "PrayerTimesFragment")
                .commit();
    }
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/activities/PrayerTimesActivity.kt
package com.mehmetakiftutuncu.muezzin.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import com.mehmetakiftutuncu.muezzin.R
import com.mehmetakiftutuncu.muezzin.activities.preferences.PreferencesActivity
import com.mehmetakiftutuncu.muezzin.fragments.NoPlacesFoundFragment
import com.mehmetakiftutuncu.muezzin.fragments.PrayerTimesFragment
import com.mehmetakiftutuncu.muezzin.models.Place
import com.mehmetakiftutuncu.muezzin.utilities.Pref
import com.stephentuso.welcome.WelcomeScreenHelper

class PrayerTimesActivity: MuezzinActivity() {
    private val welcomeScreenHelper: WelcomeScreenHelper by lazy {
        WelcomeScreenHelper(this, WelcomeActivity::class.java)
    }

    private val ctx: Context by lazy { this }

    private var shownWelcomeScreen = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_prayertimes)

        shownWelcomeScreen = savedInstanceState?.getBoolean("shownWelcomeScreen") ?: false

        if (!shownWelcomeScreen) {
            shownWelcomeScreen = true

            if (Pref.Application.getVersion(ctx) < 4) {
                // Installed version 2.0 for the first time
                Toast.makeText(ctx, R.string.welcome_updateNotice, Toast.LENGTH_LONG).show()
                Pref.edit(ctx) { clear() }
                welcomeScreenHelper.forceShow()
                Pref.Application.setVersion(this)
            } else {
                welcomeScreenHelper.show(savedInstanceState)
            }
        }
    }

    override fun onResume() {
        super.onResume()

        Pref.Places.getCurrentPlace(ctx)?.run { showPrayerTimes(this) } ?: showNoPlacesFound()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.settings, menu)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_preferences -> {
                startActivity(Intent(this, PreferencesActivity::class.java))
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        welcomeScreenHelper.onSaveInstanceState(outState)
        outState.putBoolean("shownWelcomeScreen", shownWelcomeScreen)

        super.onSaveInstanceState(outState)
    }

    private fun showNoPlacesFound() {
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.frameLayout_prayerTimesContainer, NoPlacesFoundFragment(), "NoPlacesFoundFragment")
            .commit()
    }

    private fun showPrayerTimes(place: Place) {
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.frameLayout_prayerTimesContainer, PrayerTimesFragment(place.toBundle()), "PrayerTimesFragment")
                .commit()
    }
}